### PR TITLE
storage/copy-to-s3: Parquet format support for nested types: array, list, map

### PR DESCRIPTION
**Pull request by rjobanp**
### Motivation

Part of the work towards copy-to-s3 private-preview: https://github.com/MaterializeInc/materialize/issues/7256

### Tips for reviewer

Map and List types are handled in a fairly straightforward way, but Array types are tricky due to their nature in PG/MZ.
I've chosen to represent them as a Struct with two fields - one containing the array elements and the other containing the number of dimensions in the array.

There were also some slight API differences between the StructBuilder, ListBuilder, and MapBuilder as compared to the primitive-type builders. If you'd like to learn more about those, see:
https://docs.rs/arrow/latest/arrow/array/builder/struct.StructBuilder.html
https://docs.rs/arrow/latest/arrow/array/builder/struct.MapBuilder.html
https://docs.rs/arrow/latest/arrow/array/builder/struct.GenericListBuilder.html

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/arrow-util/src/builder.rs [hunk 1]
```diff
@@ -228,11 +229,178 @@ fn scalar_to_arrow_datatype(scalar_type: &ScalarType) -> Result<(DataType, Strin
...
+            ColBuilder::StringBuilder(StringBuilder::with_capacity(item_capacity, data_capacity))
+        }
+        DataType::LargeUtf8 => ColBuilder::LargeStringBuilder(LargeStringBuilder::with_capacity(
+            item_capacity,
+            data_capacity,
+        )),
+        DataType::Decimal128(precision, scale) => ColBuilder::Decimal128Builder(
+            Decimal128Builder::with_capacity(item_capacity)
+                .with_precision_and_scale(*precision, *scale)?,
+        ),
```
Comment (rjobanp) [inline src/arrow-util/src/builder.rs:342]: these lines were simply moved from the `ArrowColumn::new` method below -- the handling of `DataType::[List, Struct, Map]` right below this is new

File: src/arrow-util/src/builder.rs [hunk 2]
```diff
@@ -228,11 +229,182 @@ fn scalar_to_arrow_datatype(scalar_type: &ScalarType) -> Result<(DataType, Strin
...
         // https://github.com/apache/arrow-rs/blob/0d031cc8aa81296cb1bdfedea7a7cb4ec6aa54ea/parquet/src/arrow/arrow_writer/mod.rs#L859
         // ScalarType::Interval => DataType::Interval(arrow::datatypes::IntervalUnit::DayTime)
+        ScalarType::Array(inner) => {
+            // Postgres / MZ Arrays are weird, since they can be multi-dimensional but this is not
+            // enforced in the type system, so can change per-value.
+            // We use a struct type with two fields - one containing the array elements as a list
+            // and the other containing the number of dimensions the array represents. Since arrays
+            // are not allowed to be ragged, the number of elements in each dimension is the same.
+            let (inner_type, inner_name) = scalar_to_arrow_datatype(inner)?;
+            // TODO: Document these field names in our copy-to-s3 docs
```
Comment (ParkMyCar) [inline src/arrow-util/src/builder.rs:239]: do we have an issue to add these docs? if not can we add one?

File: test/testdrive/copy-to-s3-minio.td [hunk 3]
```diff
@@ -259,3 +259,17 @@ $ s3-verify-data bucket=copytos3 key=parquet_test/9 sort-rows=true
...
+> INSERT INTO t3 VALUES (LIST[[1.25, 2.5],[223.3333]], LIST['a', 'b'], '{a=>{b=>2.5}}', ARRAY[[1, 2],[3, 5]]);
+> INSERT INTO t3 VALUES (LIST[[0.0], NULL], NULL, NULL, NULL);
+> INSERT INTO t3 VALUES (NULL, LIST[NULL], '{a=>NULL}', ARRAY[[1, 2], [NULL, NULL]]::int[][]);
+
+> COPY t3 TO 's3://copytos3/parquet_test/10' WITH (AWS CONNECTION = aws_conn, FORMAT = 'parquet');
+
+$ s3-verify-data bucket=copytos3 key=parquet_test/10 sort-rows=true
+[[0.0], ]   // allow-trailing-whitespace
+ [] {a: } {items: [1, 2, , ], dimensions: 2}
+[[1.25, 2.5], [223.3333]] [a, b] {a: {b: 2.5}} {items: [1, 2, 3, 5], dimensions: 2}
```
Comment (ParkMyCar) [inline test/testdrive/copy-to-s3-minio.td:275]: nice!

**Review by ParkMyCar — APPROVED** [review]
This all looks great to me!

File: src/arrow-util/src/builder.rs (hunk 2 above)
Comment (rjobanp) [reply src/arrow-util/src/builder.rs:239]: yeah I've got an all-encompassing issue here https://github.com/MaterializeInc/materialize/issues/26884 that I'm working on today - I'll be documenting all the non-obvious type transformations we deal with

**Pull request merged**